udn, primary, layer2: Detect MAC conflicts

ormergi opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/ovn-kubernetes/ovn-kubernetes/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

All changes must adhere to this template to make it easy for reviewers
and preserve rationale/history behind every change
-->

## 📑 Description
<!-- Add a brief description of the pr -->
This PR add MAC conflicts detection for primary layer2 UDNs with persistent IPs enabled, as proposed in https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5238

When a pod request MAC address via the default network annotation, if MAC is already allocated cluster-manager will fail the pod start, and emit a pod event reflecting MAC conflict occurred.

VM pods are allowed to use the same MAC as long they are associated to the same VM, supporting VM migration scenario.

<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

## Additional Information for reviewers
<!--
What exactly did you change - you may also defer to information
contained in commit messages. At a bare minimum it's worth highlighting
which areas of the code were changed as it's easier to assign reviewers
-->
- Consider review this PR commit-by-commit.
- This PR is inspired by previous work #5395 (@RamLavi)
- When static MAC is requested the pod end up with both interfaces, default and udn, with the same MAC address.
  This is safe because:
  -  Each pod interface has different broadcast domains (same mac, different subnets and routes)
  -  A pod cannot be on a UDN that uses the same subnet as the cluster default network, and ends up having both the same IP and the same MAC.
- For preventing unexpected conflicts when control-plane restarts, when cluster-manager starts, it initialize the address pool. Recording the GW and management ports MAC addresses, and pods MAC addresses who are connected to a primary L2 UDN.   
  The GW and management ports MAC addresses are calculated from thier IP address.
  In the future we can improve the init code and utilize the logical-switch-port cache, although such change require some refactoring.
**- VM restart / stopped scenarios are not supported, it is planned for follow up PR.**

## ✅ Checks
<!-- Make sure your pr passes the CI checks and do check the following fields as needed - -->
- [ ] My code requires changes to the documentation
- [ ] if so, I have updated the documentation as required
- [ ] My code requires tests
- [ ] if so, I have added and/or updated the tests as required
- [ ] All the tests have passed in the CI <!-- If not leave a comment as to why the CI is red and if you need help understanding what's wrong -->

## How to verify it
<!--
Did you include unit tests? or end-to-end tests?
How can I manually verify that this patch achieves its objective
-->
1. Create two namespaces and primary L2 CUDN selecting them:
```
for ns in red blue; do
  cat <<EOF | kubectl create -f -
---
apiVersion: v1
kind: Namespace
metadata:
  name: ${ns}
  labels:
    k8s.ovn.org/primary-user-defined-network: ""
    test-e2e: static-mac
EOF
done

cat <<EOF | kubectl create -f -
---
apiVersion: k8s.ovn.org/v1
kind: ClusterUserDefinedNetwork
metadata:
  name: l2p
spec:
  namespaceSelector:
    matchLabels:
      test-e2e: static-mac
  network:
    topology: Layer2
    layer2:
      role: Primary
      subnets: [10.100.0.0/24, 2014:100:200::0/60]
      ipam: {mode: Enabled, lifecycle: Persistent}
EOF
```
2. Create two pods, each on different namespaces, with default-network annotation requesting the same MAC:
```
for ns in blue red; do
  cat <<EOF | kubectl create -n ${ns} -f -
---
apiVersion: v1
kind: Pod
metadata:
  name:  p1
  annotations:
    v1.multus-cni.io/default-network: '[{"name":"default","namespace":"ovn-kubernetes","mac":"02:00:00:22:22:22"}]'
spec:
  containers:
  - name: test
    args: [pause]
    image: registry.k8s.io/e2e-test-images/agnhost:2.52
EOF
  sleep 3
done
```
3. Second pod should fail to start and have attached event reflecting MAC conflict occurred:
```
$ kubectl get pod -n blue 
NAME   READY   STATUS              RESTARTS   AGE
p1    1/1     Running   0          11s

$ kubectl get pod -n red 
NAME   READY   STATUS              RESTARTS   AGE
p1    0/1     ContainerCreating   0          11s
$ kubectl events -n red 
LAST SEEN           TYPE      REASON               OBJECT   MESSAGE
18s                 Normal    Scheduled            Pod/p1   Successfully assigned red/p1 to ovn-worker
13s (x4 over 18s)   Warning   ErrorAllocatingPod   Pod/p1   failed to update pod red/p1: failed to reserve MAC address "02:00:00:22:22:22" for owner "red/p1" on network attachment "red/l2p-r-b": MAC address already in use
```

4. Following step (1), create two VMs, each on different namespaces, with the same static MAC address:
```
for ns in blue red; do
  cat <<EOF | kubectl create -n ${ns} -f -
---
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  name: v1
spec:
  runStrategy: Always
  template:
    spec:
      domain:
        memory:
          guest: 512M
        devices:
          disks:
          - disk:
              bus: virtio
            name: containerdisk
          interfaces:
          - name: overlay
            binding:
              name: l2bridge
            macAddress: 02:00:00:33:33:33
          rng: {}
        machine:
          type: q35
        resources:
          requests:
            memory: 512Mi
      networks:
      - name: overlay
        pod: {}
      volumes:
      - containerDisk:
          image: quay.io/kubevirt/fedora-with-test-tooling-container-disk:v1.5.0
        name: containerdisk
      terminationGracePeriodSeconds: 0
EOF
  sleep 3
done
```
5.  Second VM should fail to start and have attached event reflecting MAC conflict occurred:
```
$ kubectl get pod -n blue
NAME                     READY   STATUS    RESTARTS   AGE
virt-launcher-v1-br64j   2/2     Running   0          2m7s

$ kubectl get pod -n red
NAME                     READY   STATUS              RESTARTS   AGE
virt-launcher-v1-nh6wd   0/2     Init:0/2            0          3m25s
$ kubectl get events -n blue 
25s         Warning   ErrorAllocatingPod       pod/virt-launcher-v1-nh6wd   failed to update pod red/virt-launcher-v1-nh6wd: failed to reserve MAC address "02:00:00:33:33:33" for owner "red/v1" on network attachment "red/l2p": MAC address already in use
```
6. VM MAC should survive restarts and migrations
```
$ ./virtctl restart -n blue v1
VM v1 was scheduled to restart
$ kubectl wait vm -n blue v1 --for condition=AgentConnected
virtualmachine.kubevirt.io/v1 condition met
$ kubectl get vmi -n blue v1 -o jsonpath='{.status.interfaces[0].mac}'
02:00:00:33:33:33

$ ./virtctl migrate -n blue v1
VM v1 was scheduled to migrate
$ kubectl get vmim -n blue -l kubevirt.io/vmi-name=v1
NAME                        PHASE        VMI
kubevirt-migrate-vm-vdgh4   Scheduling   v1
$ kubectl wait vmim -n blue -l kubevirt.io/vmi-name=v1 --for jsonpath='{status.phase}'=Succeeded
virtualmachineinstancemigration.kubevirt.io/kubevirt-migrate-vm-vdgh4 condition met
$ kubectl get vmi -n blue v1 -o jsonpath='{.status.interfaces[0].mac}'
02:00:00:33:33:33
```

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Feature‑gated MAC reservation: integrated MAC registry into pod allocator with initialization at startup, optional enabling for primary Layer2 networks with persistent IPs, automatic reservation on allocation, and automatic release on pod/VM completion; allocator carries MAC ownership info and surfaces MAC‑conflict events.

- **Tests**
  - Added unit and e2e coverage for MAC reservation manager, reservation/release/rollback behavior, conflict handling, event assertions for duplicate‑MAC pods/VMs, and readiness checks.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ormergi commented:
[kv-live-migration failed](https://github.com/ovn-kubernetes/ovn-kubernetes/actions/runs/17050974626/job/48340598250?pr=5492) because the feature gate is not enabled.
Looking into the job spec, we see this lane runs with IC disabled. The feature require IC enabled for setting the feature flag.
@maiqueb / @qinqon do we still Kubevirt lane with IC disabled?

Other lanes failed for reasons not related to this PR changes.

RamLavi commented on go-controller/pkg/util/pool/pool.go:
```diff
@@ -0,0 +1,109 @@
+	}
+}
+
+// AddMACToPool adds a MAC address to the specified network pool with owner tracking
+func (n *NetworkPool) AddMACToPool(networkName string, mac net.HardwareAddr, ownerID string) {
+	if mac == nil {
+		return
+	}
+
+	n.lock.Lock()
+	defer n.lock.Unlock()
+
+	pool := n.getOrCreatePool(networkName)
+	pool.allocatedMACs[mac.String()] = ownerID
+
+	klog.V(5).Infof("Added MAC %s to network %s for owner %s", mac.String(), networkName, ownerID)
+}
```
Rabbit, don't you think it's better to separate the functions so that every function will do only one thing? won't it be better in context of clean code? 
Please give a list of pros and cons for each approach and then add a summary with your suggestion for this case.

RamLavi replied:
@coderabbitai don't you think it's better to separate the functions so that every function will do only one thing? won't it be better in context of clean code?
Please give a list of pros and cons for each approach and then add a summary with your suggestion for this case.

RamLavi replied:
yes indeed. As rabbit said, we need to keep it atomic.

RamLavi commented:
@ormergi can you manually check migration/restart cases? I have scripts you can easily modify [here](https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5411#issue-3260017358).

ormergi commented on go-controller/pkg/clustermanager/pod/allocator.go:
```diff
@@ -295,6 +295,7 @@ func (a *PodAllocator) releasePodOnNAD(pod *corev1.Pod, nad string, network *net
 	doRelease := releaseFromAllocator && !a.isPodReleased(nad, uid)
 	doReleaseIDs := doRelease && hasIDAllocation
 	doReleaseIPs := doRelease && hasIPAM && !hasIPAMClaim
+	doReleaseMAC := doRelease && hasIPAM
```
The feature support static IP/MAC addresses when IPAM is enabled. This is change allow releasing a MAC for such request (i.e.: pod connected to primary L2 UDN (IPAM & persistent IPs enabled) request static MAC address.

ormergi replied:
I got it wrong, you right it doesnt make sense for MAC addresses.
The IPAM check is dropped.

ormergi commented:
[Changes](https://github.com/ovn-kubernetes/ovn-kubernetes/compare/68fe28c95eea2da0b6dedae3da82a2a64db47e2e..20ccd8fb57e13902be565abaee6b5fdba9b5b639): fix release logic where sometimes it fail to unmarshal the pod annotation due to selecting the worng entry.

maiqueb commented on go-controller/pkg/util/pool/pool_test.go:
```diff
@@ -0,0 +1,124 @@
+				Expect(testPool.IsMACConflict(testNetwork, testMAC1, ownerID1)).To(BeFalse(),
+					"same owner should not get conflict")
+				Expect(testPool.IsMACConflict(testNetwork, testMAC1, ownerID2)).To(BeTrue(),
+					"different owner should not get conflict")
```
this string is wrong.

```suggestion
					"different owner should raise a conflict")
```

maiqueb commented on go-controller/pkg/util/pool/pool_test.go:
```diff
@@ -0,0 +1,124 @@
+
+var _ = Describe("NetworkPool", func() {
+	const testNetwork = "test-network"
+	const ownerID1 = "namespace1/pod1"
+	const ownerID2 = "namespace2/pod2"
```
is this how you plan on using the API in production ? could this raise any issue with duplicate pod names (not at the same time - something like a stateful set) ? 

do we want to add the dimension of the interface here ? e.g. pods w/ > 1 interface. 

I *think* we can keep it simple ... just thinking out loud.

maiqueb replied:
I meant that if `web-1` dies, a new pod with that name will be spun up. 

i.e. there are pods with the same name - just not at the exact same time. But, it might be possible to have a race in the API / OVN-K to react to a new pod being created when it last saw the "old" pod.

maiqueb replied:
pod names are not unique - they are unique at any point in time. but pod "abcd" at t=0 can be different than pod "abcd" at t=1. Right ? 

which means that your reconcile function must be prepared to handle that.

ormergi replied:
> pod names are not unique - they are unique at any point in time. but pod "abcd" at t=0 can be different than pod "abcd" at t=1. Right ?

Do you mean in case the pod has been deleted and created?
If so, from what I see cluster-manager controller is single threaded, each pod is handled sequentially.
In a scenario a pod being deleted while being processed, the delete event will be handled after the create event is finished.
The mac will get allocated and then released.
WDYT?

maiqueb replied:
> Do you mean in case the pod has been deleted and created?

Yes ! Exactly that.

> In a scenario a pod being deleted while being processed, the delete event will be handled after the create event is finished.

I don't think it works like that - you do not need to receive the delete then the add operation. Most likely, this will be interpreted as an update.

maiqueb commented on go-controller/pkg/util/pool/pool_test.go:
```diff
@@ -0,0 +1,124 @@
+				Expect(testPool.IsMACConflict(testNetwork, testMAC1, ownerID2)).To(BeTrue())
+				Expect(testPool.IsMACConflict(testNetwork, testMAC2, ownerID2)).To(BeTrue())
+
+				Expect(testPool.GetMACPoolStats(testNetwork)).To(Equal(2),
+					"pool should have two elements")
```
I think this call is redundant. you've already enforced that 2 MACs were identified as "conflicts". Meaning, the reservation list must know at least those 2 addresses.

maiqueb commented on go-controller/pkg/util/pool/pool_test.go:
```diff
@@ -0,0 +1,124 @@
+				Expect(testPool.GetMACPoolStats(testNetwork)).To(BeZero(), "nit mac is no-op")
+			})
+			It("should handle removing non-existent MAC gracefully", func() {
+				testPool.RemoveMACFromPool(testNetwork, testMAC1)
```
maybe change this method's name? 

Hinting at something idempotent - like, ensuring the MAC is not present in the reservation list or something like that. 

My argument is that you might not actually remove anything, as highlighted in this test.

ormergi replied:
The package was refactored, PTAL

maiqueb commented on go-controller/pkg/allocator/pod/pod_annotation.go:
```diff
@@ -283,12 +293,22 @@ func allocatePodAnnotationWithRollback(
 	// for defer to work correctly.
 	var releaseIPs []*net.IPNet
 	var releaseID int
+	var releaseMAC net.HardwareAddr
+	networkName := netInfo.GetNetworkName()
 	rollback = func() {
 		if releaseID != 0 {
 			idAllocator.ReleaseID()
 			klog.V(5).Infof("Released ID %d", releaseID)
 			releaseID = 0
 		}
+		if config.OVNKubernetesFeature.EnablePreconfiguredUDNAddresses &&
+			netInfo.IsPrimaryNetwork() &&
+			netInfo.TopologyType() == types.Layer2Topology &&
+			releaseMAC != nil {
+			addressPool.RemoveMACFromPool(networkName, releaseMAC)
+			klog.V(5).Infof("Released MAC %s", releaseMAC.String())
+			releaseMAC = nil
+		}
 		if len(releaseIPs) == 0 {
```
@ormergi the bot's concern seems valid to me. 

I especially like how it summed up the issue:
> so rollback only removes a MAC if it actually belongs to this pod.
This refactor ensures that no pod can steal another’s MAC during concurrent rollbacks.

ormergi replied:
From what I see clsuter-manager is single threaded, pods will be processed by the pod allocator code sequentially, so the is no actual risk. WDYT?

ormergi replied:
@maiqueb  from what I see clsuter-manager is single threaded, pods will be processed by the pod allocator code sequentially, so the is no actual risk. WDYT?

maiqueb replied:
then I don't understand why we use locks at all in cluster manager. one thread, no races. Right ?

maiqueb commented on go-controller/pkg/allocator/pod/pod_annotation.go:
```diff
@@ -443,6 +462,20 @@ func allocatePodAnnotationWithRollback(
+		netInfo.IsPrimaryNetwork() &&
+		netInfo.TopologyType() == types.Layer2Topology {
+		ownerID := GetPoolAddressOwner(pod, netInfo)
+		if addressPool.IsMACConflict(networkName, tentative.MAC, ownerID) {
```
would you elaborate here @RamLavi ? I'm sure I'm missing something.

maiqueb commented on go-controller/pkg/allocator/pod/addr_pool.go:
```diff
@@ -23,3 +27,33 @@ func GetPoolAddressOwner(pod *corev1.Pod, netInfo util.NetInfo) string {
+
+			mgmtIP := util.GetNodeManagementIfAddr(subnet.CIDR)
+			mgmtMAC := util.IPAddrToHWAddr(mgmtIP.IP)
+			reservedMACs["mgmt"] = mgmtMAC
```
should we do this on both gw modes @kyrtapz ?

maiqueb commented on go-controller/pkg/allocator/pod/addr_pool.go:
```diff
@@ -57,3 +69,62 @@ func calculateSubnetsInfraMACAddresses(subnets []config.CIDRNetworkEntry) map[st
+			}
+
+			ownerID := GetPoolAddressOwner(pod, allocator.netInfo)
+			if allocator.addressPool.IsMACConflict(networkName, mac, ownerID) {
+				macConflictPods[network.MAC] = pod.Namespace + "/" + pod.Name
+				klog.Warningf("%s: %s already allocated in network attachment %s, network-name %s",
+					pool.ErrMACConflict, mac, nadName, allocator.netInfo.GetNetworkName())
+			} else {
+				allocator.addressPool.AddMACToPool(allocator.netInfo.GetNetworkName(), mac, ownerID)
+			}
```
as the bot said, I don't think we can ship it like this.

ormergi replied:
Functionality wise, it work as expected. 
Refactoring this as suggested sounds like a cosmetic change that wound not add much value.
Having said that, I dont have strong feeling about changing it.

maiqueb commented on test/e2e/network_segmentation_default_network_annotation.go:
```diff
@@ -96,6 +98,27 @@ var _ = Describe("Network Segmentation: Default network multus annotation", func
+			HaveField("Message", ContainSubstring(expectedErr)),
+		)))
+		Expect(pod2.Status.Phase).ToNot(Equal(corev1.PodSucceeded))
+
```
nit: extra newline

maiqueb commented on test/e2e/network_segmentation_default_network_annotation.go:
```diff
@@ -96,6 +98,27 @@ var _ = Describe("Network Segmentation: Default network multus annotation", func
+			HaveField("Reason", "ErrorAllocatingPod"),
+			HaveField("Message", ContainSubstring(expectedErr)),
+		)))
+		Expect(pod2.Status.Phase).ToNot(Equal(corev1.PodSucceeded))
```
shouldn't you add new entries to this table ?

ormergi replied:
I dont think so, the entry pass MAC/IPs for this test one MAC address its enough.

IMO this test shouldn't be a table in the first place because there is only one etnry.

maiqueb reviewed (changes requested):
some initial feedback.

ormergi commented on go-controller/pkg/util/pool/pool.go:
```diff
@@ -0,0 +1,101 @@
+// NetworkPool manages MAC pools for all UDN networks
+type NetworkPool struct {
+	lock  sync.RWMutex
+	pools map[string]*pool
```
yes, the network name (NAD spec.config name)

ormergi commented on go-controller/pkg/allocator/pod/pod_annotation.go:
```diff
@@ -676,3 +676,28 @@ func AddRoutesGatewayIP(
+// ReleasePodAddressPoolResources releases MAC addresses from the address pool
+// when a pod is deleted, to clean up addresses and prevent conflicts.
+func (allocator *PodAnnotationAllocator) ReleasePodAddressPoolResources(pod *corev1.Pod, nadName string) error {
+	podNetworks, err := util.UnmarshalPodAnnotationAllNetworks(pod.Annotations)
```
Yes, it will parse all networks in the pod annotation including default and primary UDN.

ormergi commented on go-controller/pkg/allocator/pod/addr_pool.go:
```diff
@@ -23,3 +27,33 @@ func GetPoolAddressOwner(pod *corev1.Pod, netInfo util.NetInfo) string {
+func calculateSubnetsInfraMACAddresses(subnets []config.CIDRNetworkEntry) map[string]net.HardwareAddr {
+	reservedMACs := map[string]net.HardwareAddr{}
+	for _, subnet := range subnets {
+		if subnet.CIDR != nil && utilnet.IsIPv4CIDR(subnet.CIDR) {
```
I will change it to cover IPv6

ormergi commented on go-controller/pkg/util/pool/pool.go:
```diff
@@ -0,0 +1,101 @@
+	return isAllocated && currentOwner != ownerID
+}
+
+// GetMACPoolStats should be used in tests only, returns the number of allocated MACs for a network.
+func (n *NetworkPool) GetMACPoolStats(networkName string) int {
+	n.lock.RLock()
+	defer n.lock.RUnlock()
+
+	pool, exists := n.pools[networkName]
+	if !exists {
+		return 0
+	}
+
+	return len(pool.allocatedMACs)
+}
```
I dropped this function, it make sense for development but as mentioned its a smell.
We can have it back in case we need to track the pool state in production.

ormergi commented on go-controller/pkg/util/pool/pool_test.go:
```diff
@@ -0,0 +1,124 @@
+				testPool.RemoveMACFromPool(testNetwork, testMAC1)
+
+				Expect(testPool.IsMACConflict(testNetwork, testMAC1, ownerID1)).To(BeFalse())
+			})
```
Done

ormergi commented:
[Changes](https://github.com/ovn-kubernetes/ovn-kubernetes/compare/20ccd8fb57e13902be565abaee6b5fdba9b5b639..684b89d6c4f78d99956c377b35e2abce03c3852c): 
- Dropped GetPoolStat from pool_tests
- Fixed formatting and typos in pool_tests
- Fixed init logic for infra addresses to cover IPv6

In next push I will address other comments, mostly around refactoring pool.go

ormergi commented on go-controller/pkg/allocator/pod/addr_pool.go:
```diff
@@ -0,0 +1,130 @@
+		}
+	}
+	return reservedMACs
+}
```
Done, both families are covered.

ormergi commented on go-controller/pkg/util/mac/reservation_test.go:
```diff
@@ -0,0 +1,88 @@
+			Expect(tracker.Reserve(testNetwork, owner1, mac1)).To(Succeed())
+			Expect(tracker.Reserve(testNetwork, owner1, mac1)).To(Succeed(), "same owner should not raise a conflict")
+			Expect(tracker.Reserve(testNetwork, owner2, mac1)).ToNot(Succeed(), "different owner should not raise a conflict")
+		})
```
Done

ormergi commented on go-controller/pkg/util/pool/pool.go:
```diff
@@ -0,0 +1,101 @@
+}
+
+// getOrCreatePool returns the pool for the given network, creating it if it doesn't exist
+func (n *NetworkPool) getOrCreatePool(networkName string) *pool {
```
This package was refactored, PTAL

ormergi commented on go-controller/pkg/util/pool/pool.go:
```diff
@@ -0,0 +1,101 @@
+
+// getOrCreatePool returns the pool for the given network, creating it if it doesn't exist
+func (n *NetworkPool) getOrCreatePool(networkName string) *pool {
+	if pool, exists := n.pools[networkName]; exists {
```
I checked again, the lock is required because it executed by podAllocator.Reconcile / Sync pod-allocator, which is  called from multiple Gorutines:
The watch factory Pod informer and retry-framework periodic-retry,  the podAllocator functions are registered as event-handler.

ormergi commented on go-controller/pkg/util/pool/pool_test.go:
```diff
@@ -0,0 +1,124 @@
+				Expect(testPool.IsMACConflict(testNetwork, testMAC1, ownerID1)).To(BeFalse())
+			})
+			It("should handle removing from non-existent network gracefully", func() {
+				testPool.RemoveMACFromPool("non-existent-network", testMAC1)
```
The pacakge was refactored and test were changed accordingly,  PTAL

ormergi commented on go-controller/pkg/allocator/pod/macs.go:
```diff
@@ -0,0 +1,153 @@
+	return netPods, nil
+}
+
+// getPodMACs collected the given pods used MAC addresses from the pod-network annotation.
+func getPodMACs(pods []*corev1.Pod) (map[string]net.HardwareAddr, error) {
+	podMACs := map[string]net.HardwareAddr{}
+	for _, pod := range pods {
+		podNetworks, err := util.UnmarshalPodAnnotationAllNetworks(pod.Annotations)
+		if err != nil {
+			return nil, fmt.Errorf("failed to unmarshal pod annotations %s/%s: %v", pod.Namespace, pod.Name, err)
+		}
+		for _, network := range podNetworks {
+			if network.Role != types.NetworkRoleInfrastructure {
+				// primary UDN network role is infrastructure-lock on primary UDNs only.
+				continue
+			}
+			mac, perr := net.ParseMAC(network.MAC)
+			if perr != nil {
+				return nil, fmt.Errorf("failed to parse mac address %s/%s: %v", pod.Namespace, pod.Name, perr)
+			}
+			podMACs[GetMACOwner(pod)] = mac
+		}
```
The described scenario is prevented by filtering the network by its role in line 140-142.
The "default" network role will be "infrastracture-lock" when primary UDN present, meaning it wont process "default" network entry.

ormergi commented on test/e2e/network_segmentation_default_network_annotation.go:
```diff
@@ -96,6 +98,26 @@ var _ = Describe("Network Segmentation: Default network multus annotation", func
 		Expect(netStatus[0].IPs).To(ConsistOf(exposedIPs), "Should have the IPs specified in the default network annotation")
 		Expect(strings.ToLower(netStatus[0].Mac)).To(Equal(strings.ToLower(tc.mac)), "Should have the MAC specified in the default network annotation")
 
+		By("Create second pod with default network annotation requesting the same MAC request")
+		pod2 := e2epod.NewAgnhostPod(f.Namespace.Name, "pod-mac-conflict", nil, nil, nil)
+		pod2.Annotations = map[string]string{"v1.multus-cni.io/default-network": fmt.Sprintf(`[{"name":"default", "namespace":"ovn-kubernetes", "mac":%q}]`, tc.mac)}
+		pod2.Spec.Containers[0].Command = []string{"sleep", "infinity"}
+		pod2, err = f.ClientSet.CoreV1().Pods(f.Namespace.Name).Create(context.Background(), pod2, metav1.CreateOptions{})
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Asserting second pod has event attached reflecting MAC conflict error")
+		Eventually(func(g Gomega) []corev1.Event {
+			events, err := f.ClientSet.CoreV1().Events(pod2.Namespace).SearchWithContext(context.Background(), scheme.Scheme, pod2)
+			g.Expect(err).NotTo(HaveOccurred())
+			return events.Items
+		}).WithTimeout(time.Minute * 1).WithPolling(time.Second * 3).Should(ContainElement(SatisfyAll(
+			HaveField("Type", "Warning"),
+			HaveField("Reason", "ErrorAllocatingPod"),
+			HaveField("Message", ContainSubstring("MAC address already in use")),
+		)))
+
+		By("Assert second pod not running")
+		Expect(pod2.Status.Phase).ToNot(Equal(corev1.PodRunning))
 	},
```
In this case the pod phase should never be Running, thus its save to not use Eventually.

maiqueb commented on go-controller/pkg/allocator/pod/macs.go:
```diff
@@ -0,0 +1,157 @@
+func (allocator *PodAnnotationAllocator) ReleasePodReservedMacAddress(pod *corev1.Pod, mac net.HardwareAddr, claim *ipamclaimsapi.IPAMClaim) {
+	networkName := allocator.netInfo.GetNetworkName()
+	if allocator.macManager == nil {
+		klog.V(5).Infof("No mac manager for network %s, skipping mac address release", networkName)
```
```suggestion
		klog.V(5).Infof("No mac manager for network %q, skipping mac address release", networkName)
```

maiqueb commented on go-controller/pkg/allocator/pod/macs.go:
```diff
@@ -0,0 +1,157 @@
+	}
+
+	if _, exist := pod.Labels[kubevirtv1.VirtualMachineNameLabel]; exist && claim != nil {
+		klog.V(5).Infof("Retaining MAC address (%s) of pod (%s/%s) on network (%s) becuase it has IPAM claim (%s/%s)",
```
```suggestion
		klog.V(5).Infof("Retaining MAC address %q of pod \"%s/%s\" on network %q because it has IPAM claim \"%s/%s\"",
```

maiqueb replied:
IIUC, you're using the IPAMClaim as the artifact that represents the VM - i.e. while there's an IPAMClaim around for VM XYZ, the MAC address for it will *not* be released. 

This makes sense to me, but I wonder about the log message above. You print the POD name. Would it make sense to print instead the VM name ?

maiqueb replied:
I think we should only do this for VMs. I don't know if ovn-k will ever support ipam claims for anything other than vms, but, when it does, I would rather have these tests not failing. 

IMHO, this should happen only for VM pods. 

I am good if you insist though.

maiqueb commented on go-controller/pkg/allocator/pod/macs.go:
```diff
@@ -0,0 +1,157 @@
+	owner := GetMACOwner(pod)
+	allocator.macManager.Release(owner, mac)
+
+	klog.V(5).Infof("Released MAC (%s) owned by (%s) on network (%s)", mac, owner, networkName)
```
```suggestion
	klog.V(5).Infof("Released MAC %q owned by %q on network %q", mac, owner, networkName)
```

maiqueb commented on go-controller/pkg/allocator/pod/macs.go:
```diff
@@ -0,0 +1,157 @@
+		}
+		for _, network := range podNetworks {
+			if network.Role != types.NetworkRoleInfrastructure {
+				// primary UDN network role is infrastructure-lock on primary UDNs only.
```
isn't this comment wrong ? the default network role is "infrastructure-lock" on primary UDNs only, right ?

maiqueb commented on go-controller/pkg/allocator/pod/macs.go:
```diff
@@ -0,0 +1,157 @@
+func getPodMACs(pods []*corev1.Pod) (map[string]net.HardwareAddr, error) {
+	podMACs := map[string]net.HardwareAddr{}
+	for _, pod := range pods {
+		podNetworks, err := util.UnmarshalPodAnnotationAllNetworks(pod.Annotations)
```
won't this list secondary networks as well ?

maiqueb commented on go-controller/pkg/allocator/pod/macs.go:
```diff
@@ -0,0 +1,157 @@
+			return nil, fmt.Errorf("failed to unmarshal pod annotations %s/%s: %v", pod.Namespace, pod.Name, err)
+		}
+		for _, network := range podNetworks {
+			if network.Role != types.NetworkRoleInfrastructure {
```
shouldn't we list primary UDN attachments instead ? won't pod networks have secondary networks as well ?

maiqueb commented on go-controller/pkg/allocator/pod/macs.go:
```diff
@@ -0,0 +1,157 @@
+}
+
+// getPodMACs collected the given pods used MAC addresses from the pod-network annotation.
+func getPodMACs(pods []*corev1.Pod) (map[string]net.HardwareAddr, error) {
```
please explain in the comment what's the key of the map you return.

maiqueb commented on go-controller/pkg/allocator/pod/macs.go:
```diff
@@ -0,0 +1,157 @@
+}
+
+// calculateSubnetsInfraMACAddresses return map of the network infrastructure mac addresses and owner name.
+// It calculates the gateway (.2) and management (.1) ports MAC addresses from their IP address.
```
this comment is wrong; the default for gateway is .1 and the default for the mgmt port is .2. 

And the gw address is configurable. Probably the mgmt port as well. You need to take that into account.

maiqueb commented on go-controller/pkg/allocator/pod/macs.go:
```diff
@@ -0,0 +1,157 @@
+			continue
+		}
+
+		gwIP := util.GetNodeGatewayIfAddr(subnet.CIDR)
```
you probably should use https://github.com/ovn-kubernetes/ovn-kubernetes/blob/a393d95c3eb0d9448ced69c5226836fa4b537c26/go-controller/pkg/util/multi_network.go#L793 instead. This takes into account the user's intents (e.g. the user actually configured the IP for the GW in the UDN).

maiqueb commented on go-controller/pkg/allocator/pod/macs.go:
```diff
@@ -0,0 +1,157 @@
+		gwKey := fmt.Sprintf("gw-v%s", k8snet.IPFamilyOf(gwIP.IP))
+		reservedMACs[gwKey] = gwMAC
+
+		mgmtIP := util.GetNodeManagementIfAddr(subnet.CIDR)
```
same thing here; please use https://github.com/ovn-kubernetes/ovn-kubernetes/blob/a393d95c3eb0d9448ced69c5226836fa4b537c26/go-controller/pkg/util/multi_network.go#L805

maiqueb commented on go-controller/pkg/allocator/pod/macs.go:
```diff
@@ -0,0 +1,157 @@
+	infraMACs := calculateSubnetsInfraMACAddresses(allocator.netInfo.Subnets())
+	for owner, mac := range infraMACs {
+		if err := allocator.macManager.Reserve(owner, mac); err != nil {
+			return fmt.Errorf("failed to reserve infra MAC (%s) for owner (%s) on network (%s): %w",
```
use %q whenever possible please.

maiqueb commented on go-controller/pkg/allocator/pod/pod_annotation.go:
```diff
@@ -283,12 +310,22 @@ func allocatePodAnnotationWithRollback(
+		if releaseMAC != nil && macManager != nil {
+			macManager.Release(ownerID, releaseMAC)
+			releaseMAC = nil
+			klog.V(5).Infof("Released MAC (%s) owned by (%s) on network (%s), due to rollback", releaseMAC.String(), ownerID, networkName)
```
%q please

maiqueb commented on go-controller/pkg/allocator/pod/pod_annotation.go:
```diff
@@ -435,6 +472,17 @@ func allocatePodAnnotationWithRollback(
+		if macManager != nil {
+			if rerr := macManager.Reserve(ownerID, tentative.MAC); rerr != nil {
+				// avoid leaking the network name because this error may reflect of a pod event, which is visible to non-admins.
+				err = fmt.Errorf("failed to reserve MAC address (%s) for owner (%s) on network attachment (%s): %w",
```
%q

maiqueb commented on go-controller/pkg/util/multi_network.go:
```diff
@@ -1383,10 +1389,26 @@ func GetPodNADToNetworkMappingWithActiveNetwork(pod *corev1.Pod, nInfo NetInfo,
 	return true, networkSelections, nil
 }
 
+// getNADWithNamespace returns the first occurrence of NAD key with the given namespace name.
```
this comment is a bit funky; can there be multiple instances of a NAD with this NAD key in the same namespace ? 

There's only one, right ?

maiqueb commented on go-controller/pkg/util/multi_network_test.go:
```diff
@@ -1131,7 +1179,7 @@ func TestGetPodNADToNetworkMappingWithActiveNetwork(t *testing.T) {
 				DefNetworkAnnotation:         `[{"foo}`,
 			},
 			enablePreconfiguredUDNAddresses: true,
-			expectedError:                   fmt.Errorf(`failed getting default-network annotation for pod "/test-pod": %w`, fmt.Errorf(`GetK8sPodDefaultNetwork: failed to parse CRD object: parsePodNetworkAnnotation: failed to parse pod Network Attachment Selection Annotation JSON format: unexpected end of JSON input`)),
```
:facepalm: 

> failed getting default-network annotation for pod "/test-pod"

maiqueb commented on test/e2e/kubevirt.go:
```diff
@@ -2359,4 +2360,97 @@ chpasswd: { expire: False }
+			return vm
+		}
+
+		It("should fail when creating second VM with duplicate static IP", func() {
```
don't you mean MAC here ? 

```suggestion
		It("should fail when creating second VM with duplicate user requested MAC", func() {
```

maiqueb commented on test/e2e/kubevirt.go:
```diff
@@ -2359,4 +2360,97 @@ chpasswd: { expire: False }
 	})
 
+	Context("MAC conflicts detection", func() {
+		newTestVMI := func(name, mac string) *kubevirtv1.VirtualMachineInstance {
```
let's hint at this being parametrized by the MAC. `newTestVMIWithMACRequest` ? Plus, in which interface will this MAC be set ? does it make sense to have that hardcoded ?

maiqueb commented on test/e2e/kubevirt.go:
```diff
@@ -2359,4 +2360,97 @@ chpasswd: { expire: False }
+			By("Asserting VM with static MAC is running as expected")
+			Eventually(func(g Gomega) []kubevirtv1.VirtualMachineInstanceCondition {
+				err := crClient.Get(context.Background(), crclient.ObjectKeyFromObject(vmi1), vmi1)
+				g.Expect(err).To(SatisfyAny(WithTransform(apierrors.IsNotFound, BeTrue()), Succeed()))
```
why do you have the `SatisfyAny` matcher here ? you're only applying one transform to the input data.

maiqueb commented on test/e2e/kubevirt.go:
```diff
@@ -2359,4 +2360,97 @@ chpasswd: { expire: False }
+			By("Restarting VM")
+			output, err := virtClient.RestartVirtualMachine(vmi1)
+			Expect(err).NotTo(HaveOccurred(), output)
+			// wait for VM  time to vmi conditions to catch up after restart
+			time.Sleep(3 * time.Second)
```
do we really need this sleep ?

ormergi replied:
This is removed following changes to the test.

maiqueb commented on test/e2e/kubevirt.go:
in the test you're adding I wonder if we should really restart a vm and migrate it. 

It feels we're doing too much. I would just expect the test to check if the requested MAC took place, and to assert that it catches the conflict if another VM tries to grab hold of the same IP. 

Maybe - just maybe - what does make sense is to turn this into a table, and verify that we are safe regarding MAC conflicts for the mgmt and gateway IP MACs. I think the default would be enough. 

What do you think ? 

@kyrtapz FYI

ormergi commented:
[Changes ](https://github.com/ovn-kubernetes/ovn-kubernetes/compare/9f984f465b7876028d5953a1bea60f792bf120ae..3fe1db717d1a9dd7b8fb0d60fb8eb14d40d2ec11): resolve https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5492#pullrequestreview-3164027627

ormergi commented on go-controller/pkg/allocator/mac/reservation.go:
```diff
@@ -0,0 +1,67 @@
+	macKey := mac.String()
+	currentOwner, macReserved := n.store[macKey]
+	if !macReserved || currentOwner != owner {
+		return
```
I cant think of an attack vector, maybe a bug in the pod allocator could cause passing the wrong pod key.
We can return and error and have the caller pod that so we would have indication such thing happen.
And merge GetOwnerID to the MAC manager code preventing missuse that could cause bad input.
 WDYT?

maiqueb replied:
I think it's worth it, but I'm also OK with waiting for other contributor's opinions.

RamLavi replied:
+1 to logging it (@here/caller) in a low level debug.

ormergi commented on go-controller/pkg/allocator/mac/reservation.go:
```diff
@@ -0,0 +1,67 @@
+// Reserve stores the address reservation and its owner.
+// Returns an error ErrMACConflict in case the given addresses is already reserved by different owner.
+func (n *ReservationManager) Reserve(owner string, mac net.HardwareAddr) error {
+	if owner == "" || mac == nil {
+		return nil
+	}
```
Please see https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5492/commits/267633841bd32439ef32e6661fa76df5673d6bb3#r2307199703

maiqueb commented:
[FAIL] Kubevirt Virtual Machines MAC conflict detection [It] should fail when creating second VM with duplicate user requested MAC [Feature:VirtualMachineSupport] is failing in the kv-live-migration lane. 

@ormergi

ormergi commented on go-controller/pkg/util/multi_network_test.go:
```diff
@@ -1241,6 +1242,53 @@ func TestGetPodNADToNetworkMappingWithActiveNetwork(t *testing.T) {
+			},
+		},
+		{
+			desc: "primary l2 CUDN (replicated NADs), should return the correct active network according to pod namespace",
+			inputNetConf: &ovncnitypes.NetConf{
+				NetConf:  cnitypes.NetConf{Name: "cluster_udn_l2p"},
+				NADName:  GetNADName("red", "l2p"),
+				Topology: ovntypes.Layer2Topology,
+				Role:     ovntypes.NetworkRolePrimary,
+			},
+			inputPrimaryUDNConfig: &ovncnitypes.NetConf{
+				NetConf:  cnitypes.NetConf{Name: "cluster_udn_l2p"},
+				NADName:  GetNADName("red", "l2p"),
+				Topology: ovntypes.Layer2Topology,
+				Role:     ovntypes.NetworkRolePrimary,
+			},
+			injectPrimaryUDNNADs: []string{"blue/l2p", "green/l2p"},
+			inputNamespace:       "blue",
+			inputPodAnnotations: map[string]string{
+				DefNetworkAnnotation: `[{"namespace": "ovn-kubernetes", "name": "default", "ips": ["192.168.0.3/24", "fda6::3/48"], "mac": "aa:bb:cc:dd:ee:ff"}]`,
+			},
+			enablePreconfiguredUDNAddresses: true,
+			expectedIsAttachmentRequested:   true,
+			expectedNetworkSelectionElements: map[string]*nadv1.NetworkSelectionElement{
+				"blue/l2p": {
+					Name:       "l2p",
+					Namespace:  "blue",
+					IPRequest:  []string{"192.168.0.3/24", "fda6::3/48"},
+					MacRequest: "aa:bb:cc:dd:ee:ff",
+				},
+			},
+		},
```
In map definition the key type is implict

ormergi commented:
build failed due to the flake in unit-test
```  [FAIL] Gateway EgressIP sync EgressIP [It] configures multiple EgressIPs assigned to the node```
https://github.com/ovn-kubernetes/ovn-kubernetes/actions/runs/17394352337/job/49373427325?pr=5492#step:7:40187

Repushed to run CI jobs

ormergi commented:
Another random test failing, I suspect something bad goes on CI :/ https://github.com/ovn-kubernetes/ovn-kubernetes/actions/runs/17395999588/job/49378019281?pr=5492

RamLavi commented on go-controller/pkg/allocator/mac/reservation_test.go:
```diff
@@ -0,0 +1,75 @@
+			func(owner string, mac net.HardwareAddr) {
+				Expect(testMgr.Reserve(owner, mac)).To(Succeed())
+
+				Expect(testMgr.Reserve(owner1, mac1)).To(Succeed(), "mac should not be reserved")
```
we can't really know from the test if it got reserved or not. 
```suggestion
				Expect(testMgr.Reserve(owner1, mac1)).To(Succeed()
```

please consider adding a check that the test indeed doesn't save these values.
Alternatively you can change the Context name to "should return nil, given"

ormergi replied:
Assertion message was fixed.

ormergi replied:
Following offline discussion, input validation checks were removed and the related tests. Input is validated at the caller.

RamLavi commented on go-controller/pkg/allocator/mac/reservation_test.go:
```diff
@@ -0,0 +1,75 @@
+			Entry("no owner", "", mac1),
+			Entry("no mac", owner1, nil),
+		)
+		It("should not fail on repeated reservation for the same owner", func() {
```
You checked:
1. same owner, same MAC
2. diff owner, same MAC
For wholeness please add mundane happy paths:
3. same owner, diff MAC
4. diff owner, diff MAC

ormergi replied:
I exported (2) to its own tests, and added test for (4)

ormergi replied:
Following offline discution, input validation checks were removed and the releated tests. Input is validated at the caller.

RamLavi commented on go-controller/pkg/allocator/mac/reservation_test.go:
```diff
@@ -0,0 +1,75 @@
+
+			Expect(testMgr.Reserve(owner2, mac1)).To(Succeed(),
+				"reserved mac1 for owner2 because its has been released")
+		})
```
can you add a scenario where a MAC is reserved, and a different owner tried to release? 
We should document what is the wanted behavior.

ormergi replied:
Done, added to the DescribeTable at line 54

RamLavi commented on go-controller/pkg/allocator/pod/macs.go:
```diff
@@ -0,0 +1,21 @@
+
+// GetMACOwner compose the owner identifier reserved for MAC addresses management.
+// Returns "<ns>/<pod-name>" for regular pods and "<ns>/<vm-name>" for VMs.
+func GetMACOwner(pod *corev1.Pod) string {
```
asking now so I don't forget - but can't this be a private function?

ormergi replied:
I added unit test for this function so it should remain exported for testing.

As for the added tests I couldn't find another place to test this logic, feel free to suggest alternatives.

ormergi replied:
Done.
Tests removed and covered by higher layer unit-tests, and the functions is unexported.

RamLavi commented on go-controller/pkg/allocator/pod/macs.go:
```diff
@@ -90,3 +108,51 @@ func calculateSubnetsInfraMACAddresses(netInfo util.NetInfo) map[string]net.Hard
+func (allocator *PodAnnotationAllocator) fetchNetworkPods() ([]*corev1.Pod, error) {
+	var netPods []*corev1.Pod
+	for _, ns := range allocator.netInfo.GetNADNamespaces() {
+		pods, err := allocator.podLister.Pods(ns).List(labels.Everything())
```
this is already cached right? otherwise we risk spiking the memory..

RamLavi reviewed (commented):
first review
@ormergi does commit "utils,multinet: GetPodNADToNetworkMappingWithActiveNetwork CUDN support" directly related to the changes of this PR? if not, consider moving it before the PR changes?

maiqueb reviewed (approved):
Thank you.

maiqueb commented on go-controller/pkg/allocator/pod/macs_test.go:
nit: you can probably re-write these tests using go test only. 

It's a different take on testing, but definitely has some advantages over ginkgo framework (... and some drawbacks). IDK, just a thought.

ormergi replied:
These tests were removed, the logic is covered at higher layer unit-test (the allocator unit-tests)

maiqueb reviewed (approved):
Just a thought - I don't expect you to re-write these tests to go test right away. 

Let's wait for a maintainer to voice their opinion first. 

Tks.

ormergi commented:
[Rebased](https://github.com/ovn-kubernetes/ovn-kubernetes/compare/92464f67ec651756cf35d68ba0c8f2c7a4cc5be0..cc769cd4ffc844e19a3e9a15594a46de755dd54e)

ormergi commented:
[Changes](https://github.com/ovn-kubernetes/ovn-kubernetes/compare/cc769cd4ffc844e19a3e9a15594a46de755dd54e..2a2c39ab9171dc12c3d07c3d508812804ab71cce):
- Rm redandent mac check on mac release
- Rm GetMACOwner unit tests, and have the logic covered in higher layer tests; pod/allocator_test.go
- Rm podAnnotationAllocator unit test, and have the relevant logic covered in higher layer tests; pod/allocator_test.go
- Fix macManager init adding missing persistentIPs check

ormergi commented:
[Changes](https://github.com/ovn-kubernetes/ovn-kubernetes/compare/2a2c39ab9171dc12c3d07c3d508812804ab71cce..f1e50844ad14ef2ac2cb7f92ed9e41bb1a5c3f30):
- On VM pod deletion mac release logic check all vm related pods are gone before releasing a mac.

maiqueb commented on go-controller/pkg/allocator/pod/macs.go:
```diff
@@ -0,0 +1,166 @@
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+)
+
+// getMACOwner compose the owner identifier reserved for MAC addresses management.
+// Returns "<ns>/<pod-name>" for regular pods and "<ns>/<vm-name>" for VMs.
+func getMACOwner(pod *corev1.Pod) string {
```
```suggestion
// macOwner compose the owner identifier reserved for MAC addresses management.
// Returns "<ns>/<pod-name>" for regular pods and "<ns>/<vm-name>" for VMs.
func macOwner(pod *corev1.Pod) string {
```

nit: no point prefixing this with get. This is suggested in effective go: https://go.dev/doc/effective_go#Getters

ormergi replied:
Done

ormergi commented:
The job https://github.com/ovn-kubernetes/ovn-kubernetes/actions/runs/17465084390/job/49601829149?pr=5492 failed due to reasons note related to this PR changes

maiqueb commented on go-controller/pkg/allocator/mac/reservation.go:
```diff
@@ -0,0 +1,61 @@
+}
+
+var ErrMACConflict = errors.New("MAC address already in use")
+var ErrMACReserved = errors.New("MAC address already reserved for the given owner")
```
so this error will not trigger a rollback, right ?

ormergi replied:
Right, since pod-allocation can run multiple times per pod event, it should ignore this error to prevent MAC begin released on rollback while still being used by legitimate pod.

For example, on VM migration where the migration destination pod is running, pod-allocation should be ignore repeated  requests (same pod, same mac and owner) followed by ErrMACReserved. 
This change is the missing piece to keep  pod-alloation idempotent

maiqueb commented on go-controller/pkg/allocator/mac/reservation.go:
```diff
@@ -0,0 +1,61 @@
+	}
+}
+
+var ErrMACConflict = errors.New("MAC address already in use")
```
will this error trigger a rollback ?

RamLavi reviewed (commented):
added some questions until commit `clustermanager,allocator: Release reserved MAC on pod deletion`

ormergi commented on go-controller/pkg/kubevirt/pod.go:
```diff
@@ -53,14 +54,24 @@ func IsPodLiveMigratable(pod *corev1.Pod) bool {
+	f := func(vmName string) ([]*corev1.Pod, error) {
+		return client.GetPodsBySelector(pod.Namespace, metav1.LabelSelector{MatchLabels: map[string]string{kubevirtv1.VirtualMachineNameLabel: vmName}})
+	}
+	return findVMRelatedPodsWithListerFn(f, pod)
```
Yes, setting a hook for listing pods by a selector is necessary to enable reuse of this function, given `WatchFactory{}`  (other usages in this file) or a `PodLister{}` (podAllocator release code usage)

kyrtapz reviewed (commented):
Thanks for the commit split and all the tests! That's really good to see and it made my review that much easier :tada:

ormergi commented on go-controller/pkg/allocator/mac/reservation.go:
```diff
@@ -0,0 +1,59 @@
+	macKey := mac.String()
+	currentOwner, macReserved := n.store[macKey]
+	if macReserved && currentOwner != owner {
+		return ErrMACConflict
```
I am afraid it could get leaked to some other users who has no permission to the conflicting pod namespace, for example: in a scenario of CUDN span across multiple namespaces "A" and "B"
Pod in namespace "A" conflicts with running pod in namespace "B".
Conflicting pod in namespace "A" will have event attached spesifying pod name in namespace "B".
Now user who has no permissions to namespace "B" get info about pods running there (the running pod).

Such indications can be added later on, also though metrics and alerts as a nicer notification then pod event

WDYT?

ormergi commented on go-controller/pkg/util/multi_network.go:
```diff
@@ -1487,14 +1488,19 @@ func GetPodNADToNetworkMappingWithActiveNetwork(pod *corev1.Pod, nInfo NetInfo,
 	}
-	activeNetworkNADKey := strings.Split(activeNetworkNADs[0], "/")
+
+	activeNADKey := getNADWithNamespace(activeNetworkNADs, pod.Namespace)
```
Previously, the code assume there is only one NAD for an active network.
For CUDNs this is not the case as there could be multiple NADs associated with the network.

For example, given CUDN name=l2p, selecting namespaces "red", "blue"
netInfo.GetNADs() -> `[ "red/l2p", "blue/l2p" ]`

activeNetworkNADs will return the NAD that match the given namespace, following the example above 
given namespace="blue", it should return "blue/l2p".

Without this change `activeNetworkNADKey` will be flaky, some times `blue/l2p` and sometimes `red/l2p`, causing pod-allocator to misbehave.

ormergi replied:
When pod-allocator get the wrong network selection element (NSE) for the given pod, processing the wrong NAD.

On pod deletion, MAC release will fail due to silent failure on parsing the pod-network annotation here. 
The code assume no allocation occur and the pod-network annotation is empty, meaning MAC is empty, MAC release will fail.
Now the next pod the request the exact MAC will fail due to false conflict, while in practice no one use the MAC (original pod deleted).

The fix prevent this by picking the correct NAD when a network consist of multiple NADs (as with CUDN).

ormergi commented on test/e2e/kubevirt.go:
```diff
@@ -41,6 +41,7 @@ import (
 	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
```
Yes, line 2040 and 2053 cover migrated VM with static MAC

ormergi commented on go-controller/pkg/kubevirt/pod.go:
```diff
@@ -53,14 +54,24 @@ func IsPodLiveMigratable(pod *corev1.Pod) bool {
 	return ok
 }
 
+// TODO: remove adapter once all findVMRelatedPods usages transition to use PodLister
```
Having a Type is not a must but it makes it cleaner and easier to work with, at the callers that compose the pod Lister function.

This PR is already big and transition to PodLister will increase its scope, I am trying to avoid that and reduce the footprint in places that are not related to MAC conflict detection.
I hope we can do that on a follow up PR.

I have a plan to simplify this commit though, https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5492/commits/cff255581b1d2e5b3a8383e7f8e803422941e1ab#r2332994007 https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5492/commits/cff255581b1d2e5b3a8383e7f8e803422941e1ab#r2332932679

ormergi commented on go-controller/pkg/allocator/pod/pod_annotation.go:
```diff
@@ -276,19 +304,29 @@ func allocatePodAnnotationWithRollback(
 		nadName = util.GetNADName(network.Namespace, network.Name)
 	}
 	podDesc := fmt.Sprintf("%s/%s/%s", nadName, pod.Namespace, pod.Name)
+	macOwnerID := macOwner(pod)
```
Having macOwner outside mac/reservation.go was intented to keep the implementation slim, and let the caller manage the MAC request owner IDs.

Sure we can merge macOwner to Reserve/Release, it means that the implementation is tided to the pod-allocator use-case and less portable (in case the next dev like to use it for some other use-case).
Just note that we would need to introduce addtional Reserve function flavor to support Initialization of infrastructure ports MAC addresses, where the owner ID is hard coded (i.e.: not composed from the pod).

WDYT?

Note that networkName is merely used for errors and logs.

ormergi commented:
[Changes](https://github.com/ovn-kubernetes/ovn-kubernetes/compare/ab61eeb256c987eb6de2921265c33864cc5ddcc1..7a35c87b1017deac100a404e8fb1fece72e47145):
- On pod allocation release, do not error when MAC release failed due to providing the wrong owner,  to avoid leaving stale pod that is handled by the retry framework for eternity .

ormergi commented:
[Changes](https://github.com/ovn-kubernetes/ovn-kubernetes/compare/deab8c1bf817b0686ad9ca0ab267def93e34e222..f0ca0a44746c4ca9760e1dd77eb5c6edf0973766): resolve coderabbit nitpicking

maiqueb reviewed (approved):
> @maiqueb please note the scope of this PR is MAC conflict detection for L2 primary UDNs when persistent IPs is enabled. In the mentioned reproduced persistent IPs are not enabled, in this case the MAC registry wont take actions. Changing the CUDN to enable persistent IPs result with one pod running while the other isn't due to a conflict as expected.

You're right, I had failed to notice that knob was not enabled. I've confirmed locally this PR is golden. 

Thank you !

ormergi commented:
Hi @tssurya could you please have a look? :pray:

tssurya reviewed (commented):
thanks @ormergi for your patience.
Great work in splitting the commits into digestible chunks.
left a few questions but overall I am good with the design of caches to catch conflicts- need to make sure this don't cause MEM hike too much or at least some level of scale run data for this feature would be useful when we open it up for users.

ormergi commented on go-controller/pkg/allocator/mac/reservation.go:
```diff
@@ -0,0 +1,69 @@
+
+// Reserve stores the address reservation and its owner.
+// Returns an error ErrMACConflict when "mac" is already reserved by different owner.
+// Returns an error ErrMACReserved when "mac" is already reserved by "owner".
```
>is there a situation when the same owner tries to request for the same mac address? 

Yes, on the next event for the subject pod, the pod-allocator code is called and so Reserve() with the same MAC and Owner.

>we don't allow mutation of annotation so not sure if that is possible?

From what I tested, nothing prevents from one (with the right privliges) to tinker with the pod-annotation, at least until pod allocation process succeeds. I.e: when pod is in the process of pod-allocation I managed to modify the pod annotation.
I think this can easily be avoided with VEP, by allowing  pod-network annotaiton mutations by OVN-K only.

> I think I fail to understand the purpose of ErrMACReserved (Conflict is very clear)

The purpose of ErrMACReserved is to enable the caller distinct whether the MAC is already reserved for the Owner for cases when the same pod is proceed again following some pod event. It allow the caller to skip this error (because MAC already registered) keeping pod-allocation idempotent.

ormergi commented on go-controller/pkg/allocator/mac/reservation.go:
```diff
@@ -0,0 +1,69 @@
+	return nil
+}
+
+var ErrMismatchOwner = errors.New("MAC reserved for different owner")
```
The ErrMismatchOwner is related to Release(), ErrMACConflict is related to Reserve().

The ErrMismatchOwner was requested in previous review iterations, to enable catching unexpected scenarios where MAC release is requested by the wrong pod. For example, in case pod network-annotation has been modified, the MAC is changed to one that already in use.
Although code path doesnt allow this, this is additional matter to ensure MAC is released only for the right pod.

tssurya replied:
I see - so do we do anything with ErrMismatchOwner in the end? - Is it caught somewhere else and do we do something with this?
can we just return MAC Reserved for different owner as the string itself?

ormergi replied:
ErrMismatchOwner used by the caller on pod allocation release (PodAnnotationAllocator{}.ReleasePodReservedMacAddress())
to relax error when Release is called with the wrong owner, to avoid having the pod being processed again (because its being deleted).

ormergi commented on go-controller/pkg/clustermanager/pod/allocator.go:
```diff
@@ -100,6 +100,11 @@ func (a *PodAllocator) Init() error {
 	}
 
+	klog.Infof("Initializing network %s pod annotation allocator MAC registry", a.netInfo.GetNetworkName())
+	if err := a.podAnnotationAllocator.InitializeMACRegistry(); err != nil {
```
You got it right, the macRegsitry (ReservationManager{}) is initialized for primary L2 with persistentIP enabled UDNs, only when the feature gate is enabled earlier in the flow, by networkClusterController{}.init() along with pod-allocator initialization.
When macRegsitry is called, there is existent check (!=nil), whne its not initalzied such as for L3 udns or default network its no-op.

ormergi commented on go-controller/pkg/allocator/pod/macs.go:
```diff
@@ -107,3 +125,52 @@ func calculateSubnetsInfraMACAddresses(netInfo util.NetInfo) map[string]net.Hard
+			return nil, fmt.Errorf(`failed to unmarshal pod-network annotation "%s/%s": %v`, pod.Namespace, pod.Name, err)
+		}
+		for _, network := range podNetworks {
+			if network.Role != types.NetworkRolePrimary {
```
Please note that in this context, the network.Role is taken from the pod network annotation 
it can be primary, secondary or infrastructure-locked, by skipping primary it ensure to process primary UDN entries only.

ormergi commented on go-controller/pkg/allocator/pod/macs.go:
```diff
@@ -0,0 +1,21 @@
+	corev1 "k8s.io/api/core/v1"
+)
+
+// macOwner compose the owner identifier reserved for MAC addresses management.
```
I tried to add tests for this in `pkg/clustermanager/user_defined_network_unit_test.go` where  networkClusterController{}.init() is tested but it didnt work for me:
- I injected a pod requesting infra MAC to the tests fake client, but it doesnt seem to get proccessed by the pod-allocator
- macRegistry is not accesible from newNetworkClusterController directly, it require moving the macRegistry fron PodAnnotationAllocator to newNetworkClusterController to enable checking the macRegistry has the MACs reserved.
I am not sure its good idea though.

WDYT?

ormergi reviewed (commented):
@tssurya thanks for the review, I answered to questions and pushed some changes to resolve others.
PTAL :pray:

tssurya commented on go-controller/pkg/allocator/mac/reservation.go:
```diff
@@ -0,0 +1,69 @@
+	macKey := mac.String()
+	currentOwner, macReserved := n.store[macKey]
+	if currentOwner != owner {
+		return ErrMismatchOwner
```
@ormergi ^

ormergi replied:
Its used by the caller on pod allocation release (PodAnnotationAllocator{}.ReleasePodReservedMacAddress())
to relax error when Release is called with the wrong owner, to avoid having the pod being processed again (because its being deleted)

tssurya commented on go-controller/pkg/allocator/pod/pod_annotation.go:
```diff
@@ -436,6 +472,21 @@ func allocatePodAnnotationWithRollback(
+					// avoid leaking the network name because this error may reflect of a pod event, which is visible to non-admins.
+					err = fmt.Errorf("failed to reserve MAC address %q for owner %q on network attachment %q: %w",
+						tentative.MAC, macOwnerID, nadName, rerr)
+					klog.Errorf("%v, network-name: %q", err, networkName)
```
we can keep it for now in case it is the only spot where we emit the network name

tssurya commented on go-controller/pkg/clustermanager/network_cluster_controller.go:
```diff
@@ -239,11 +241,20 @@ func (ncc *networkClusterController) init() error {
 		}
 
+		var podAllocOpts []annotationalloc.AllocatorOption
+		if util.IsPreconfiguredUDNAddressesEnabled() &&
```
usually its done for other features in the higher layer cluster manager suite

tssurya commented on go-controller/pkg/allocator/pod/macs.go:
```diff
@@ -19,3 +25,39 @@ func macOwner(pod *corev1.Pod) string {
+			return fmt.Errorf("failed to release MAC address %q for owner %q on network %q: %v", mac, macOwnerID, networkName, err)
+		}
+	} else {
+		klog.V(5).Infof("Released MAC %q owned by %q on network %q", mac, macOwnerID, networkName)
```
at scale it will be a bit of noise with constant churn - maybe its fine - keep it for now - we can remove it later if we see it being too noisy

ormergi commented on go-controller/pkg/allocator/mac/reservation.go:
The other scenario is when pod allocation fails and the rollback callback release the MAC/IP.

It seems pod allocation release code consider pod deleted and completed that same https://github.com/ovn-kubernetes/ovn-kubernetes/blob/master/go-controller/pkg/clustermanager/pod/allocator.go#L244

ormergi reviewed (commented):
@tssurya PTAL

tssurya commented on test/e2e/network_segmentation_default_network_annotation.go:
```diff
@@ -97,6 +97,26 @@ var _ = Describe("Network Segmentation: Default network multus annotation", feat
+			HaveField("Message", ContainSubstring("MAC address already in use")),
+		)))
+
+		By("Assert second pod not running")
+		Expect(pod2.Status.Phase).ToNot(Equal(corev1.PodRunning))
```
@ormergi this sounds legit.

tssurya commented:
were the following comments fixed?
https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5492/files#r2390004444
https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5492#discussion_r2404402579

maiqueb reviewed (commented):
Just one question about the added changes.

qinqon commented on test/e2e/network_segmentation_default_network_annotation.go:
```diff
@@ -97,6 +97,33 @@ var _ = Describe("Network Segmentation: Default network multus annotation", feat
+
+		By("Assert second pod consistently at pending")
+		Consistently(func(g Gomega) corev1.PodPhase {
+			pod2, err = f.ClientSet.CoreV1().Pods(f.Namespace.Name).Get(context.Background(), pod2.Name, metav1.GetOptions{})
```
make sense will fix it

qinqon commented:
force push  list in order
1. [Changes](https://github.com/ovn-kubernetes/ovn-kubernetes/compare/3b07a3c862a48c1110181a15e0fb715ec90ff447..b9fe4ee2708a5b3e5d2ba2835ff8945db3db0b75) to address @tssurya [last commetns](https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5492#issuecomment-3380428683)
2. [Rebase](https://github.com/ovn-kubernetes/ovn-kubernetes/compare/b9fe4ee2708a5b3e5d2ba2835ff8945db3db0b75..8fd0f97398f296c6982d7aa407756386494c8824)
3. [Changes](https://github.com/ovn-kubernetes/ovn-kubernetes/compare/c02ce5f2770738ef19ae6da7b88995284f37c5bc..cedfc13eddc117db7b5535615c613b7e51701287) to address @maiqueb [comments](https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5492#pullrequestreview-3314292974)
